src/evm.zig

```diff
@@ -448,25 +449,31 @@ pub fn Evm(config: EvmConfig) type {
                 return CallResult.failure(self.getCallArenaAllocator(), 0) catch unreachable;
             };
 
-            const call_gas = params.getGas();
-            if (!config.disable_gas_checks) {
-                if (call_gas == 0) return CallResult.failure(self.getCallArenaAllocator(), 0) catch unreachable;
+            const gas_limit = params.getGas();
+            if (comptime !config.disable_gas_checks) {
+                if (gas_limit == 0) return CallResult.failure(self.getCallArenaAllocator(), 0) catch unreachable;
             }
 
-            const initial_gas = call_gas;
-
-            const execution_gas = blk: {
-                const GasConstants = primitives.GasConstants;
-                const intrinsic_gas = switch (params) {
-                    .create, .create2 => GasConstants.TxGasContractCreation, // 53000 for contract creation
-                    else => GasConstants.TxGas, // 21000 for regular calls
-                };
-
-                // Check if we have enough gas for intrinsic cost
-                if (params.getGas() < intrinsic_gas) return CallResult.failure(self.getCallArenaAllocator(), 0) catch unreachable;
-
-                break :blk params.getGas() - intrinsic_gas;
-            };
+            // Calculate all base gas costs:
+            // - intrinsic gas (base 21,000 or 53,000 for create)
+            // - calldata gas (from zero and non-zero bytes, non-zero being more expensive)
+            // - floor gas (same as above, will enfore a minimum gas spent at the end of the tx)
+            // TODO: cost for access list accounts & storage (EIP-2930)
+            // gas += access_list_accounts * ACCESS_LIST_ADDRESS (2400);
+            // gas += access_list_storages * ACCESS_LIST_STORAGE_KEY (1900);
+            // TODO: cost for authorization list (EIP-7702)
+            // gas += authorization_list_num * PER_EMPTY_ACCOUNT_COST (25000);
+            const calldata_tokens = config.eips.tokens_in_calldata(params.getInput());
+            const intrinsic_gas_cost = config.eips.intrinsic_gas_cost(params.isCreate());
+            const calldata_gas_cost = config.eips.calldata_gas_cost(calldata_tokens, params.isCreate(), params.getInput().len);
+            const floor_gas_cost = config.eips.floor_gas_cost(calldata_tokens);
+
+            const initial_gas_cost = intrinsic_gas_cost + calldata_gas_cost;
+            if (comptime !config.disable_gas_checks) {
+                if (gas_limit < initial_gas_cost) return CallResult.failure(self.getCallArenaAllocator(), 0) catch unreachable;
+                if (gas_limit < floor_gas_cost) return CallResult.failure(self.getCallArenaAllocator(), 0) catch unreachable;
```

Comment:
This doesn't look right to me we still need to deal with not underflowing. Might be a good idea to remove disable gas cost as an option right now seems to be randomly implemented without much thought and too much vibes everywhere